### PR TITLE
[WIP] [Magiclysm] Bugfix infinite whiskey

### LINKED ISSUE #48643: Endless Flask Provides Truly endless quantities of Whiskey
# Describe the bug

Endless Flask should regenerate whiskey in 7 charges over 24 hours.   However, instead you activate it once wait some number of hours, less than 12 I believe,  and then you are able to activate it at will to produce 7 charges of whiskey.

# Steps To Reproduce
Create a world with Magiclysm.  
Create a character.
Debug spawn in endless whiskey flask.  
Activate see it say "hasn't finished..."
Wait til midnight
Activate Flask.  Activate it time after time.  More than 7 times and it keeps refilling.

# Expected behavior
Many different possibilities of how it could work.  But this is not one.

# Screenshots
Unlimited Flask

# Versions and configuration
- OS: Windows
    - OS Version: 10.0 2004
- Game Version: 0.E-10282-gf22beb8 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Bionic Professions [package_bionic_professions],
    Magiclysm [magiclysm]
]

# Additional context

### PR DESCRIPTION
**Pull request by Maleclypse**
#### Summary
Bugfixes "Endless Whiskey Flask shouldn't be instantly endless"

#### Purpose of change
Fixes #48643

#### Describe the solution
Removes the the activation action.  Replaced with ammunition data to make whiskey an ammo that can regenerate.  All of the above works.  However the PC can no longer drink directly from the flask.  Pouring whiskey from Flask works fine though and can be drank then. 

#### Describe alternatives you've considered
Obsoleting Endless Flask until exploits can be fixed.

#### Testing
whiskey

### CODE REVIEW

**Comment by Salty-Panda** [comment]
Have you considered removing transform_age and adding charges_per_use to item?
With charges being `mana`, not whiskey.

Also, your current solution only regens one charge of whiskey per day.

BTW, I have some experience with use_action transform, I believe that time counter for transform age is copied directly to target item, so it might be bug working the same way as #47887, where after transformation your new filled item inherits properties of old one, since only itype is changed.

File: data/mods/Magiclysm/items/enchanted_misc.json [hunk 1]
```diff
@@ -115,17 +115,7 @@
     "price_postapoc": "850 USD",
     "category": "clothing",
     "relic_data": { "charge_info": { "recharge_type": "periodic", "time": "24 h", "regenerate_ammo": true } },
```
Comment (Maleclypse) [inline data/mods/Magiclysm/items/enchanted_misc.json:117]:
```suggestion
    "relic_data": { "charge_info": { "recharge_type": "periodic", "time": "4 h", "regenerate_ammo": true } },
```

**Comment by Maleclypse** [comment]
> Have you considered changing delayed_transform to transform, removing transform_age and adding charges_per_use to item?
> With charges being `mana`, not whiskey.
> 
> Also, your current solution only regens one charge of whiskey per day.
> 
> BTW, I have some experience with use_action transform, I believe that time counter for transform age is copied directly to target item, so it might be bug working the same way as #47887, where after transformation your new filled item inherits properties of old one, since only itype is changed.

I've moved the regen of charges of whiskey to one every 4 hours.  I'm cool if you want to make those changes as I think either change will make this item fit it's original intent. Goal is roughly one flask worth a day.  Let me know if you put up a PR to make the other change and I'll pull down this one.

**Comment by Salty-Panda** [comment]
I tried something that _should_ work, using relic charges.
<details>
  <summary>JSON</summary>

```json
"name": { "str": "endless flask" },
    "id": "mflask_hip_whiskey",
    "type": "TOOL_ARMOR",
    "copy-from": "flask_hip",
    "price_postapoc": "850 USD",
    "category": "clothing",
    "relic_data": {
      "charge_info": {
        "recharge_type": "periodic",
        "time": "1 m",
        "max_charges": 2
      }
    },
    "charges_per_use": 1,
    "max_charges": 2,
    "pocket_data": [
      {
        "pocket_type": "CONTAINER",
        "rigid": true,
        "watertight": true,
        "max_contains_volume": "250 ml",
        "max_item_volume": "17 ml",
        "max_contains_weight": "1 kg"
      }
    ],
    "use_action": {
      "target": "whiskey",
      "msg": "You open the flask and discover it full of sweet, sweet, whiskey!",
      "container": "mflask_hip_whiskey",
      "target_charges": 7,
      "moves": 50,
      "type": "transform",
      "//": "24 hours"
    }
```
 
</details>

Regardless of what method I use and where I try to put it, I can't declare `max_charges` so that they load properly. It won't accumulate charges because in `relic::try_recharge` `charge.charges >= charge.max_charges` - both values are 0. Maybe it can't be TOOL_ARMOR, maybe it's something else, not going to investigate further.

**Comment by Maleclypse** [comment]
> I tried something that _should_ work, using relic charges.
> 
> JSON
> Regardless of what method I use and where I try to put it, I can't declare `max_charges` so that they load properly. It won't accumulate charges because in `relic::try_recharge` `charge.charges >= charge.max_charges` - both values are 0. Maybe it can't be TOOL_ARMOR, maybe it's something else, not going to investigate further.

I feel like I tried to make that same change months ago and ran into the same problem, left it for someone more talented than myself to figure it out and got the suggestion I've used on discord a few days ago.

**Comment by Maleclypse** [comment]
> I tried something that _should_ work, using relic charges.
> 
> JSON
> Regardless of what method I use and where I try to put it, I can't declare `max_charges` so that they load properly. It won't accumulate charges because in `relic::try_recharge` `charge.charges >= charge.max_charges` - both values are 0. Maybe it can't be TOOL_ARMOR, maybe it's something else, not going to investigate further.

@Salty-Panda I opened a new branch to try and test Korg's preferred change of the whiskey being "NO_UNLOAD".    I removed the delay on the transformation but I can always unload the whiskey.   

`  {
    "name": { "str": "endless flask" },
    "id": "mflask_hip_whiskey",
    "type": "GENERIC",
    "copy-from": "flask_hip",
    "price_postapoc": "850 USD",
    "category": "clothing",
    "relic_data": { "charge_info": { "recharge_type": "periodic", "time": "24 h", "regenerate_ammo": true } },
    "use_action": {
      "target": "whiskey",
      "msg": "You open the flask and discover it full of sweet, sweet, whiskey!",
      "container": "mflask_hip_whiskey",
      "target_charges": 7,
      "moves": 50,
      "type": "transform"
    },
    "flags": [ "NO_UNLOAD" ]
  },`
  
  Thoughts?

**Comment by Salty-Panda** [comment]
Two possibilities

Either no unload only works with ammo, or code that handles unloading has
special case for liquids that doesn't consider this flag.

I don't recall any no unload item that wasn't ammo container so I guess the
former.

**Comment by Salty-Panda** [comment]
We have this ugly dual relic/tool system for charges that are essentially
ammo and item can only hold one type of ammo at once and there is no way to
specify which ammo should be used by specific action.
So this might be a code fix.

**Pull request closed without merging**